File: Algorithms/Java/48 Rotate Image.java
class Solution {
    public void rotate(int[][] matrix) {
        int n = matrix.length;

        // transpose
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < i; ++j) {
                int tmp = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = tmp;
            }
        }

        // reverse each row
        for (int i = 0; i < n; ++i) {
            for (int start = 0, end = n - 1; start < end; ++start, --end) {
                int tmp = matrix[i][start];
                matrix[i][start] = matrix[i][end];
                matrix[i][end] = tmp;
            }
        }
    }
}

File: Algorithms/C/48 Rotate Image.c
void swap(int *a, int *b)
{
    int temp = *a;
    *a = *b;
    *b = temp;
}

void rotate(int** matrix, int matrixSize, int* matrixColSize){
    if (matrix == NULL || matrixSize <= 1) {
        return;
    }

    // transpose
    for (int i = 0; i < matrixSize; ++i)
        for (int j = 0; j < i; ++j) {
            swap(&matrix[i][j], &matrix[j][i]);
        }

    // reverse each row
    for (int i = 0; i < matrixSize; ++i) {
        for (int start = 0, end = matrixSize - 1; start < end; ++start, --end) {
            swap(&matrix[i][start], &matrix[i][end]);
        }
    }
}
